numpy/__init__.pyi

```diff
@@ -1455,10 +1464,77 @@ class ndarray(_ArrayOrScalarCommon, Generic[_ShapeType, _DType]):
     def __iter__(self) -> Any: ...
     def __contains__(self, key) -> bool: ...
     def __index__(self) -> int: ...
-    def __lt__(self, other: ArrayLike) -> Union[ndarray, bool_]: ...
-    def __le__(self, other: ArrayLike) -> Union[ndarray, bool_]: ...
-    def __gt__(self, other: ArrayLike) -> Union[ndarray, bool_]: ...
-    def __ge__(self, other: ArrayLike) -> Union[ndarray, bool_]: ...
+
+    # The last overload is for catching recursive objects whose
+    # nesting is too deep.
+    # The first overload is for catching `bytes` (as they are a subtype of
+    # `Sequence[int]`) and `str`. As `str` is a recusive sequence of
+    # strings, it will pass through the final overload otherwise
+
+    @overload
```

Comment:
To summarize the various overloads:
1. Overload for filtering out any `str`/`bytes`-based array-likes. This is needed as `bytes` would otherwise be recognized as a `Sequence[int]` sub-type (which is bad) and `str` would otherwise be caught by overload 6. (as `str` is a sequence of strings, which is a sequence of strings, _etc._).
2. Number-based overload
3. Timedelta-based overload
4. Datetime-based overload
5. Object-based overload. `other` is typed as `Any` here as object arrays can be extremely flexible (depending on the actual underlying objects).
6. A final overload to handle all sequences whose nesting level is too deep for the previous overloads.